cpp/src/io/utilities/type_inference.cuh

```diff
@@ -0,0 +1,282 @@
+  if (decimal_cnt > 1) return false;
+  if (exponent_cnt > 1) return false;
+  // Without the exponent or a decimal point, this is an integer, not a float
+  if (decimal_cnt == 0 && exponent_cnt == 0) return false;
+
+  // Can only have one '-' per component
+  if (dash_cnt > 1 + exponent_cnt) return false;
+
+  // If anything other than these characters is present, it's not a float
+  if (digit_cnt + decimal_cnt + dash_cnt + exponent_cnt != len) return false;
+
+  // Needs at least 1 digit, 2 if exponent is present
+  if (digit_cnt < 1 + exponent_cnt) return false;
+
+  return true;
+}
+
+/**
+ * @brief Constructs column type histogram for a given column string input `data`.
+ *
+ * @tparam ColumnStringIter Iterator type whose `value_type` is a
+ * `thrust::tuple<offset_t, length_t>`, where `offset_t` and `length_t` are of integral type and
+ * `offset_t` needs to be convertible to `std::size_t`.
+ *
+ * @param[in] options View of inference options
+ * @param[in] data JSON string input
+ * @param[in] column_strings_begin The begining of an offset-length tuple sequence
+ * @param[in] size Size of the string input
+ * @param[out] column_info Histogram of column type counters
+ */
+template <typename ColumnStringIter>
+__global__ void infer_column_type_kernel(json_parse_options_view options,
+                                         device_span<char const> data,
+                                         ColumnStringIter column_strings_begin,
+                                         std::size_t size,
+                                         cudf::io::column_type_histogram* column_info)
+{
+  for (auto idx = threadIdx.x + blockDim.x * blockIdx.x; idx < size;
+       idx += gridDim.x * blockDim.x) {
+    auto const field_offset = thrust::get<0>(*(column_strings_begin + idx));
+    auto const field_len    = thrust::get<1>(*(column_strings_begin + idx));
+    auto const field_begin  = data.begin() + field_offset;
+
+    if (cudf::detail::serialized_trie_contains(
+          options.trie_na, {field_begin, static_cast<std::size_t>(field_len)})) {
+      atomicAdd(&column_info->null_count, 1);
+      continue;
+    }
+
+    if (field_len == 0) {
+      atomicAdd(&column_info->null_count, 1);
+      continue;
+    }
+    // Handling strings
+    if (*field_begin == options.quote_char && field_begin[field_len - 1] == options.quote_char) {
+      atomicAdd(&column_info->string_count, 1);
+      continue;
+    }
+
+    uint32_t digit_count    = 0;
+    uint32_t decimal_count  = 0;
+    uint32_t slash_count    = 0;
+    uint32_t dash_count     = 0;
+    uint32_t plus_count     = 0;
+    uint32_t colon_count    = 0;
+    uint32_t exponent_count = 0;
+    uint32_t other_count    = 0;
+
+    auto const maybe_hex = (field_len > 2 && *field_begin == '0' && *(field_begin + 1) == 'x') ||
+                           (field_len > 3 && *field_begin == '-' && *(field_begin + 1) == '0' &&
+                            *(field_begin + 2) == 'x');
```

Comment:
nit:
```suggestion
    auto const maybe_hex = (field_len > 2 && field_begin[0] == '0' && field_begin[1] == 'x') ||
                           (field_len > 3 && field_begin[0] == '-' && field_begin[1] == '0' &&
                            field_begin[2] == 'x');
```

Reply:
I think `*` is preferred to dereference iterators over `[]` since not all iterators support bracket access.

Reply:
the [random_access_iterator](https://en.cppreference.com/w/cpp/iterator/random_access_iterator) concept requires both to be equivalent, and the iterators seem to be pointers anyways?